Emails: Add informative message and actions when user can't add email to the selected domain

AllTerrainDeveloper opened the pull request:
#### Proposed Changes

As shown here 1200182182542585-as-1201225902975505/f we should be more informative to the user when the user is not able to add an email subscription to the domain, when the user is not the owner of the domain. 

#### Testing Instructions

1. Transfer a domain to your site, where your user is not the owner, and there are no email subscriptions attached to it
2. go to /email/{domain}/purchase/{site}
3. Assert that you see this screen:

![image](https://user-images.githubusercontent.com/5689927/192509110-75db890b-5404-457b-ba33-d960ceaea8f8.png)

Assert the following statements:
1. The "log in" redirects you to the login page
2. The reach-out link is enabled if the domain is registered with us and has privacy enabled.
3. Each of the select buttons are disabled

Assert when clicking on reach out link that you see the following screen:
![image](https://user-images.githubusercontent.com/5689927/191758121-cb484864-96b0-4f43-8777-1eb4ecc1651a.png)

Assert that the links are logging the three different actions:
![image](https://user-images.githubusercontent.com/5689927/192751835-8da0bd1b-d40d-4630-976e-884d5b739e42.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #66650

AllTerrainDeveloper committed:
Emails: Add informative message and actions when user can't add email to the selected domain

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +91,51 @@ const EmailProvidersStackedComparison = ( {
+	const contactOwnerUrl = `https://privatewho.is/?s=${ selectedDomainName }`;
+
+	const loginUrlWithRedirectToProfessionalEmail =
+		'http://wordpress.com/log-in?redirect_to=' +
```
We'll also need the parameter `email_address=` to supply the owner's username.

AllTerrainDeveloper committed:
Emails: Disable select buttons, gather information for the domain owner and use it in the new link buttons

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +99,53 @@ const EmailProvidersStackedComparison = ( {
+
+	const contactOwnerUrl = `https://privatewho.is/?s=${ selectedDomainName }`;
+
+	const purchses = useSelector( ( state ) => getSitePurchases( state, selectedSite?.ID ) );
```
This was meant to be `purchases`? :)

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +99,53 @@ const EmailProvidersStackedComparison = ( {
+
+	const teams = data as InfiniteData< UsersData > & UsersData;
+
+	const ownerUserName = teams ? teams.users[ 0 ].login : '';
```
Since we'll need similar logic for the Email Management page as well to surface messaging to non-email-subscription owners, it's probably worthwhile to create a helper function for getting the domain owner username? It'll also improve modularity of the code.

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +99,53 @@ const EmailProvidersStackedComparison = ( {
+	const ownerUserName = teams ? teams.users[ 0 ].login : '';
+
+	const userNameUrlParam = `?email_address=${ ownerUserName }`;
+	const loginUrl = `https://wordpress.com/log-in${ ownerUserName ? userNameUrlParam : '' }`;
```
We're missing the `redirect_to` parameter that brings users to back to the Email Comparisons page after logging in as the domain owner.

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -269,6 +325,35 @@ const EmailProvidersStackedComparison = ( {
+									"If you don't have access to that account, please reach out to the domain owner {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}}",
+								{
+									components: {
+										link: <a href={ loginUrl } onClick={ () => onClickLink( 'user_login' ) } />,
```
I would suggest tracking `owner_login` to be consistent with the other clicking tracking value of `owner_contact`. 

I also noticed a behaviour where the tracking call for login link appeared to have failed. This only happens when we're linking to a new page in the same browser tab, the tracking call results in `200 OK` status when the `<a>` target is set to `_blank`. Should we be worried about this? 

<img width="1379" alt="Screen Shot 2022-09-23 at 12 10 22 PM" src="https://user-images.githubusercontent.com/104910361/192005938-542f4e79-dc6b-4e18-9ebb-f8032a9109b8.png">

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/style.scss:
```diff
@@ -30,6 +30,12 @@
+.email-providers-stacked-comparison__owner-notice {
+	margin-top: 15px;
+	margin-bottom: 15px;
+	padding: 36px 24px;
```
```suggestion
	padding: 24px 24px 4px;
```
There appears to be excessive padding here... the nested `<p>` already has a `margin-bottom: 20px`. 

<img width="1060" alt="Screen Shot 2022-09-23 at 11 25 49 AM" src="https://user-images.githubusercontent.com/104910361/192007073-2f2ea405-1e83-4672-a219-7bd68d3af79d.png">

Proposing we apply consistent padding on all sides:

<img width="975" alt="Screen Shot 2022-09-23 at 12 19 51 PM" src="https://user-images.githubusercontent.com/104910361/192007283-32ab0f5c-ca75-4a3e-895c-1b8ebb059a7c.png">

wongasy commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -269,6 +325,35 @@ const EmailProvidersStackedComparison = ( {
 
 			<>
+				{ ! currentUserCanAddEmail && (
+					<PromoCard className="email-providers-stacked-comparison__owner-notice">
```
Wondering if it makes more sense for us to rename the class to `email-providers-stacked-comparison__nonowner-notice`, since this component is shown to non-owners.

wongasy reviewed (changes requested):
Making progress 👍 Please see my inline comments! One last note that, we should be sending the property `context: non-owner` for the `calypso_email_providers_comparison_page_view` event if the user is a non-owner (see requirement from 1200182182542585-as-1201225902975505/f)

olaseni commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +99,53 @@ const EmailProvidersStackedComparison = ( {
+	const loginUrl = `https://wordpress.com/log-in${ ownerUserName ? userNameUrlParam : '' }`;
+
+	const onClickLink = ( eventType: string ) => {
+		dispatch( recordTracksEvent( `calypso_email_providers_${ eventType }_click` ) );
```
Since they are built dynamically, it will be worthwhile to document the full names of the expected events.

This will build a link in the source code to the actual events in the log in the event that a developer is debugging the origin of that event in the future.

olaseni reviewed (commented):
Thanks @wongasy for getting ahead of this! There's plenty to go on already.

AllTerrainDeveloper committed:
Emails: Amend Pull Request suggestions
 - Move logic to extract the domain user name
 - Explicit function description to ease the location of the event logged in Tracks
 - Change the style class name & padding for the notice.
 - Extract the function that builds the URL login page with email comparison and user login name

AllTerrainDeveloper committed:
Emails: Remove testing flag and revert target="_blank" change on log in track event.

AllTerrainDeveloper commented:
@wongasy The tracks event seems to be working fine for me. I've addressed every other one concern  :)

AllTerrainDeveloper commented:
I have a followup pull request incoming that will refactor the message into its own component `<EmailNonDomainOwnerMessage domain={ domain } selectedSite={ selectedSite } usePromoCard={ true }>`

fredrikekelund commented on client/my-sites/email/paths.js:
```diff
@@ -298,3 +298,31 @@ export function emailManagementInbox( siteName = null ) {
+ * @param { string } selectedDomainName The domain name to purchase email subscription
+ * @returns {string} Login Url with email_address and redirect_to parameters included in the URI
+ */
+export function loginUrlWithUserNameAndRedirectToEmailProvidersComparison(
```
Maybe we shouldn't add this as a utility function in this file. Letero doesn't own the `/log-in` slug, and to my mind, this file is a single source of truth for slugs related to email management in Calypso.

AllTerrainDeveloper replied:
That's a good point, I'll include it in the new component developed.

fredrikekelund replied:
Looks like you maybe forgot to remove this? 🙂

fredrikekelund commented on client/my-sites/email/paths.js:
```diff
@@ -298,3 +298,31 @@ export function emailManagementInbox( siteName = null ) {
+			'google-workspace'
+		)
+	);
+	const loginUrl = `https://wordpress.com/log-in${
```
```suggestion
	const loginUrl = `/log-in${
```
Should we drop the origin here? Should make things easier during development.

AllTerrainDeveloper replied:
Right!

fredrikekelund commented:
Making a note that #68156 is related to this PR.

stephanethomas commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,41 @@
+
+	const domainSubscription = purchases.filter(
+		( purchase ) => purchase.id === parseInt( domain?.subscriptionId ?? '0' )
+	)[ 0 ];
```
Not a big deal but it seems a bit strange to evaluate the identifier of the domain subscription in that loop, maybe we should extract that:

```suggestion
	const domainSubscriptionId = domain?.subscriptionId ?? '0';

	const domainSubscription = purchases.filter(
		( purchase ) => purchase.id === parseInt( domainSubscriptionId )
	)[ 0 ];
```

stephanethomas commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,41 @@
+	const domainSubscription = purchases.filter(
+		( purchase ) => purchase.id === parseInt( domain?.subscriptionId ?? '0' )
+	)[ 0 ];
+
+	const fetchOptions = {
+		search: domainSubscription?.userId,
+		search_columns: [ 'ID' ],
+	};
```
Shouldn't we return early whenever some mandatory data is not available?

```suggestion
	if ( ! domainSubscription ||| ! selectedSite ) {
		return '';
	}
	
	const fetchOptions = {
		search: domainSubscription?.userId,
		search_columns: [ 'ID' ],
	};
```

AllTerrainDeveloper replied:
I did it that way as we shouldn't make conditional "useEffects". It ends with complaining, but I will check again.

AllTerrainDeveloper replied:
That's it:

![image](https://user-images.githubusercontent.com/5689927/192333761-f643adbc-523f-4ddc-adde-712addf6f208.png)

stephanethomas replied:
Oh, that's right.

stephanethomas commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +92,47 @@ const EmailProvidersStackedComparison = ( {
+	 * This function builds the event 'calypso_email_providers_owner_contact_click'
+	 * or 'calypso_email_providers_user_login_click' depending on the event type passed as parameter.
+	 *
+	 * @param { "owner_contact" | "user_login" } eventType
```
That annotation seems superfluous given it just duplicates what we already have in the function's signature.

AllTerrainDeveloper replied:
As the key is built dynamically, I thought it was good to keep the final keys as a comment so other developers would find them easily.

stephanethomas replied:
Note I was only talking about the `@param` annotation here. Good idea mentioning the Track events in the comment above although I think it's better to avoid building them dynamically in the code (as someone could forget to update that comment 🙂).

stephanethomas commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -86,12 +92,47 @@ const EmailProvidersStackedComparison = ( {
+	 * @param { "owner_contact" | "user_login" } eventType
+	 */
+	const onClickLink = ( eventType: 'owner_contact' | 'user_login' ) => {
+		dispatch(
+			recordTracksEvent( `calypso_email_providers_${ eventType }_click`, {
+				owner_login: ownerUserName,
+			} )
+		);
```
Even if it means more code, it's usually better to avoid building event names dynamically as that makes it impossible to grok them later:

```suggestion
		let eventName = `calypso_email_providers_owner_contact_click`;
		
		if ( some logic ) {
			eventName = `calypso_email_providers_user_login_click`;
		}
		
		dispatch(
			recordTracksEvent( eventName , {
				owner_login: ownerUserName,
			} )
		);
```

AllTerrainDeveloper replied:
That was the intention of the superfluous documentation :) But I will do as you suggest :)

stephanethomas commented on client/my-sites/email/email-providers-comparison/stacked/index.tsx:
```diff
@@ -269,6 +312,35 @@ const EmailProvidersStackedComparison = ( {
+							) }
+						</p>
+					</PromoCard>
+				) }
```
Something to consider: given how long the `render()` method already is, we could extract that block into a standalone component (in the same file but defined above `EmailProvidersStackedComparison`).

AllTerrainDeveloper replied:
As this message is going to be used for the next two tasks, the next one is doing that refactor and new component :)

https://github.com/Automattic/wp-calypso/pull/68158#issuecomment-1257827364

stephanethomas commented on client/my-sites/email/paths.js:
```diff
@@ -298,3 +298,31 @@ export function emailManagementInbox( siteName = null ) {
+	);
+	const loginUrl = `https://wordpress.com/log-in${
+		ownerUserName ? userNameUrlParameter : ''
+	}&redirect_to=${ redirectUrlParameter }`;
```
If `ownerUserName` is missing this will generate a url like https://wordpress.com/log-in&redirect_to=<something>, which is missing a question mark. We should probably use the `buildQueryString()` function here.

stephanethomas reviewed (changes requested):
This generates a wrong login url, when clicking it a 404 error page is displayed:

![image](https://user-images.githubusercontent.com/594356/192320124-5f6a4ed6-1d2e-40b6-a3b5-33bad20ce76a.png)

As an aside, I think it may a good idea to mention the domain name in the copy. I tested those changes with a test account that only had one domain, and I was presented with the following page when clicking `Emails` in the sidebar:

![image](https://user-images.githubusercontent.com/594356/192319446-46eb8fa6-c440-43d2-ae81-ddbe9fa1c3d5.png)

I think the following would help clarifying which domain is concerned:

> An email solution can only be purchased by the domain owner of **example.com**.

AllTerrainDeveloper committed:
[Emails] Make the URL relative to the host.

AllTerrainDeveloper committed:
Emails: Improve the final wording, build login query correctly, extract component to do not overflow the main render.

AllTerrainDeveloper committed:
Fix linting errors

wongasy commented:
>I have a followup pull request (for a different task) incoming that will refactor the message into its own component <EmailNonDomainOwnerMessage domain={ domain } selectedSite={ selectedSite } usePromoCard={ true }>

@AllTerrainDeveloper It would make sense to do the refactoring in a separate PR if we were refactoring code that exists in `trunk`. But given that we are introducing new code in this PR anyways, might as well do the refactoring here to avoid potential regression issues that might be introduced in the refactoring.

AllTerrainDeveloper committed:
Emails: Refactor the owner notice into a new component

AllTerrainDeveloper commented:
> > I have a followup pull request (for a different task) incoming that will refactor the message into its own component <EmailNonDomainOwnerMessage domain={ domain } selectedSite={ selectedSite } usePromoCard={ true }>
> 
> @AllTerrainDeveloper It would make sense to do the refactoring in a separate PR if we were refactoring code that exists in `trunk`. But given that we are introducing new code in this PR anyways, might as well do the refactoring here to avoid potential regression issues that might be introduced in the refactoring.

Done :)

AllTerrainDeveloper committed:
Emails: Remove test flag

AllTerrainDeveloper committed:
Emails: Fixed bug related with wpcom ids

AllTerrainDeveloper committed:
Emails: Add isLoading from React Query in order to exit earlier

fredrikekelund commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,51 @@
+
+	const selectedSubscriptionId = domain?.subscriptionId ?? '0';
+
+	const domainSubscription = purchases.filter(
+		( purchase ) => purchase.id === parseInt( selectedSubscriptionId )
+	)[ 0 ];
```
```suggestion
	const domainSubscription = purchases.find(
		( purchase ) => purchase.id === parseInt( selectedSubscriptionId )
	);
```
I think we could use `find` here instead of `filter`

fredrikekelund commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	const teams = data as InfiniteData< UsersData > & UsersData;
+	const ownerUser = teams.users?.filter(
```
```suggestion
	const ownerUser = teams.users?.find(
```
Same thing there about `find`

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+		parameters ? stringify( parameters, { addQueryPrefix: true, skipNulls: true } ) : '';
+
+	const buildLoginUrl = () => {
+		const redirectUrlParameter = encodeURIComponent(
+			emailManagementPurchaseNewEmailAccount(
+				selectedSite?.slug ?? '',
+				domain?.name ?? '',
+				null,
+				'login-redirect'
+			)
+		);
```
It looks like `qs.stringify` already URI encodes parameters, so we don't have to do it here.
```suggestion
		const redirectUrlParameter = emailManagementPurchaseNewEmailAccount(
			selectedSite?.slug ?? '',
			domain?.name ?? '',
			null,
			'login-redirect'
		);
```

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+			owner_login: ownerUserName,
+		};
+
+		if ( eventType === 'owner_contact' ) {
+			dispatch( recordTracksEvent( `calypso_email_providers_owner_contact_click`, properties ) );
+
+			return;
+		} else if ( eventType === 'contact_support' ) {
+			dispatch( recordTracksEvent( `calypso_email_providers_contact_support_click`, properties ) );
+
+			return;
+		}
+
+		dispatch( recordTracksEvent( `calypso_email_providers_user_login_click`, properties ) );
```
```suggestion
		if ( eventType === 'owner_contact' ) {
			dispatch( recordTracksEvent( `calypso_email_providers_owner_contact_click`, properties ) );
		} else if ( eventType === 'contact_support' ) {
			dispatch( recordTracksEvent( `calypso_email_providers_contact_support_click`, properties ) );
		} else if ( eventType === 'user_login' ) {
			dispatch( recordTracksEvent( `calypso_email_providers_user_login_click`, properties ) );
		}
```
Nitpicking, but I think this reads better. I'm guessing that the return statements are there because ESLint maybe doesn't like final `else` statements? In any case, it seems like we could do without the `return` statements here.

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+					'Email service can only be purchased by {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}}, ' +
+						'who is the owner of %(selectedDomainName)s. ' +
+						'If you have access to that account, please {{loginLink}}log in with the account{{/loginLink}} to make a purchase. ' +
+						'Otherwise, please reach out to {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}} or {{contactSupportLink}}contact support{{/contactSupportLink}}',
```
```suggestion
						'Otherwise, please {{reachOutLink}}reach out to %(ownerUserName)s{{/reachOutLink}} or {{contactSupportLink}}contact support{{/contactSupportLink}}.',
```
Missing dot. Also, I think it would make sense to have `reachOutLink` wrap the words `reach out` as well, as it's a little bit unclear what to expect from that link otherwise.

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+					'Additional mailboxes can only be purchased by {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}}, ' +
+						'who is the owner of %(selectedDomainName)s. ' +
+						'If you have access to that account, please {{loginLink}}log in with the account{{/loginLink}} to make a purchase. ' +
+						'Otherwise, please reach out to {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}} or {{contactSupportLink}}contact support{{/contactSupportLink}}',
```
```suggestion
						'Otherwise, please reach out to {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}} or {{contactSupportLink}}contact support{{/contactSupportLink}}.',
```
A dot here too.

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+import './style.scss';
+
+type EmailNonDomainOwnerMessageProps = {
+	domain?: ResponseDomain;
+	selectedSite?: SiteDetails | null;
+	usePromoCard: boolean;
+	source: 'email-comparison' | 'email-management';
```
Nit, but maybe we should sort this alphabetically.

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+		},
+	};
+
+	function getReason() {
```
I initially thought of this as a potential render optimization (using an inline function in a React component makes it difficult for React to avoid unnecessary rerenders). However, after thinking about it, I'm not 100% sure that it actually has an impact in this case.

So this is potentially a subjective issue, although I still think it would be clearer. 

My thinking is that we ought to remove the function and instead just use a regular `if` statement to assign a value to a variable. Like this:

```tsx
let text: TranslateResult;

if ( source === 'email-comparison' ) {
    text = translate( ... );
} else if ( source === 'email-management' ) {
    text = translate( ... );
}
```

fredrikekelund commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,51 @@
+
+	const teams = data as InfiniteData< UsersData > & UsersData;
+	const ownerUser = teams.users?.filter(
+		( user ) => user.linked_user_ID === domainSubscription?.userId
```
```suggestion
		( user ) => user.ID === domainSubscription?.userId
```
In my testing, the user's ID property was `ID` rather than `linked_user_ID`. Could the property names vary, or should this be `user.ID`?

wongasy replied:
@AllTerrainDeveloper and I spent some time troubleshooting this and we found that the endpoint returns:
* `ID` for simple sites
* both `linked_user_ID` (`id_wpcom` for v2 endpoint) and `ID` for atomic sites, where `linked_user_ID` is the actual user ID and `ID` is the Jetpack external user id

Dani [updated the code](https://github.com/Automattic/wp-calypso/pull/68158/files#diff-2ea6652302a0be15504ab269dde396fa7fc4353df0e5d8fb3f47324a32955129R47) to first check for `linked_user_ID`, if it it's undefined, then fallback to `ID`.

(See related thread here: p1664277912392969/1664274185.306699-slack-C01TJT42PMM)

fredrikekelund replied:
Just making a quick note that I also saw the same thing [here](https://github.com/Automattic/wp-calypso/blob/4f40f3ff0d18598384d7a6a9c924aa929c5cb6e3/client/my-sites/domains/domain-management/transfer/transfer-to-other-user/transfer-domain-to-other-user.jsx#L37).

Might be out of scope for this PR, but perhaps it would be useful to have a utility function to retrieve the user ID?

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+		switch ( source ) {
+			case 'email-comparison':
+				return translate(
+					'Email service can only be purchased by {{reachOutLink}}%(ownerUserName)s{{/reachOutLink}}, ' +
```
```suggestion
					'Email service can only be purchased by %(ownerUserName)s, ' +
```
I think it's hard to know what to expect from `reachOutLink` in this sentence, and given that it has such "dramatic" effects (i.e. links to an unfamiliar external site), I think it's preferable to just skip it altogether here.

AllTerrainDeveloper replied:
We should discuss this with @wongasy, this is part of the implementation details and requirements.

AllTerrainDeveloper replied:
Discussed in Slack. Removing the link :)

fredrikekelund reviewed (commented):
I think we should maybe consider a loading state for `EmailNonDomainOwnerMessage`. At least when I tested sandboxed, it took a few seconds for `ownerUserName` to load, which resulted in an odd looking message (see screenshot).

![current-loading-state](https://user-images.githubusercontent.com/1101677/192535554-01bc31f6-4556-4673-8829-51e5348690ad.png)

Maybe we could interject some sort of loading component into the translated string while `ownerUserName` is loading..? Here's a rough mockup.

![loading-mockup](https://user-images.githubusercontent.com/1101677/192535922-ac0599eb-0eed-4f33-8de9-97672414e96b.png)

AllTerrainDeveloper committed:
Emails: Fallback linked user (jetpack) to wpcom user id.

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+				<a
+					rel="noopener noreferrer"
+					href={ loginUrl }
+					onClick={ () => onClickLink( 'user_login' ) }
```
Looks like we're still having issues with this link. When I tested in Firefox, the address in the address bar updated to the login link, but the login page did not load. When I added `target="_blank"` the login page opened in a new tab as expected.

<img width="1377" alt="Screen Shot 2022-09-27 at 8 07 28 PM" src="https://user-images.githubusercontent.com/104910361/192658802-396fd979-13f2-449d-b546-24fc5fc5b9dd.png">

AllTerrainDeveloper replied:
Weird issue!

AllTerrainDeveloper replied:
I found a workaround (used in other places of calypso. I think it's better to keep the same tab instead of adding a new tab.

fredrikekelund replied:
Could be something to do with `/log-in` not being defined as a section in `client/sections.js`? Maybe try with `rel="external"` on the link to force a non-push based navigation.

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+
+	const onClickLink = ( eventType: 'owner_contact' | 'user_login' | 'contact_support' ) => {
+		const properties = {
+			owner_login: ownerUserName,
```
Let's not log the username, we should not be collecting unnecessary user data.

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,139 @@
+		};
+
+		if ( eventType === 'owner_contact' ) {
+			dispatch( recordTracksEvent( `calypso_email_providers_owner_contact_click`, properties ) );
```
Instead of creating three custom events, let's create one single event `calypso_email_providers_nonowner_click` and send the action as a property `action`:
* `contact`
* `login`
* `support`

This will minimize the number of events to create. Funnel creation will also be simplified, where we just need to create one single funnel and get all click data by simply segmenting by `action`.

wongasy reviewed (changes requested):
@AllTerrainDeveloper I've added a few feedbacks and it looks like there's a number of feedback from @fredrikekelund that haven't been addressed yet :)

>Maybe we could interject some sort of loading component into the translated string while ownerUserName is loading..?

Regarding Fredrik's suggestion of adding a loading state, a simpler way to handle this might be to surface a different message when the `username` is being fetched:

>Email service can only be purchased by the owner of %(selectedDomainName)s. If you have access to that account, please log in with the account to make a purchase. Otherwise, please  {{contactSupportLink}}contact support{{/contactSupportLink}}.

Note that we should render "log in" as plain text before `username` becomes available, since we don't want to link users to the login page without pre-populating the username.

AllTerrainDeveloper committed:
Emails: Address feedback from Pull Request. Added new translations and fixed URL in login action

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,140 @@
+			action: eventType,
+		};
+
+		dispatch( recordTracksEvent( `calypso_email_providers_nonowner_click`, properties ) );
```
Nit, but the Calypso team has [recommended](https://github.com/Automattic/wp-calypso/pull/64746/files#r901446982) that we rely on `recordTracksEvent` from `calypso/lib/analytics` instead of the Redux equivalent.

AllTerrainDeveloper replied:
Than you! I wasn't aware of this recommendation!

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,140 @@
+
+	const isPrivacyAvailable = domain?.privacyAvailable;
+
+	const buildQueryString = ( parameters = {} ) =>
+		parameters ? stringify( parameters, { addQueryPrefix: true, skipNulls: true } ) : '';
```
We could move this function definition outside `EmailNonDomainOwnerMessage`, no?

fredrikekelund commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,140 @@
+
+	const translateOptions = {
+		components: {
+			loginLink: <a href={ loginUrl } onClick={ () => onClickLink( 'login' ) } />,
```
```suggestion
			loginLink: <a href={ loginUrl } onClick={ () => onClickLink( 'login' ) } rel="external" />,
```
I think it's preferable to force non-push based navigation this way instead of relying on `window.location.href = ...`. That should work better if the user `Cmd`+click's the link, for example.

fredrikekelund commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,52 @@
+	);
+	const ownerUserName = ownerUser?.login ?? '';
+
+	return ownerUserName ?? '';
```
```suggestion

	return ownerUser?.login ?? '';
```
It looks like we're defaulting to `''` twice here.

fredrikekelund commented on client/components/data/query-domain-owner-username/index.ts:
```diff
@@ -0,0 +1,52 @@
+		( purchase ) => purchase.id === parseInt( selectedSubscriptionId )
+	);
+
+	const { data, isLoading } = useUsersQuery(
+		selectedSite?.ID,
+		{},
+		{
+			enabled: domainSubscription !== undefined,
+		}
+	);
```
```suggestion
	const { data } = useQuery(
		[ 'user', selectedSite?.ID, domainSubscription?.userId ],
		() => wpcom.req.get( `/sites/${ selectedSite?.ID }/users/${ domainSubscription?.userId }` ),
		{ enabled: domainSubscription !== undefined }
	);
```
Not a requirement, but we _could_ query the API directly for a particular user and probably improve loading time. Would this work with the whole `linked_user_ID` vs `ID` question?

AllTerrainDeveloper replied:
I tried, but when querying the user when it is a jetpack user id it is not found. it gave me tons of headaches to realize.

I mean, it will work when the site is not a jetpack site, but luckily (or unluckily for my headaches), the domainSubscirption.userId is not the same in the user endpoint.

I've also tried to look for different things, using the same endpoint, and specifying other fields for searching it, with no luck (when the site is a jetpack site, as I said your solution would work flawlessly for non-jetpack sites)

fredrikekelund replied:
Just to make sure I understand: `domainSubscription.userId` can vary between a WordPress.com user ID and a Jetpack external user ID?

AllTerrainDeveloper replied:
Nop, the one that vary is the one provided by the User endpoint. Sometimes userId is correct, and sometimes is the Jetpack one.

fredrikekelund replied:
Okay, I see. So looking up `/sites/:site_id/users/:user_id` won't work with `user_id = domainSubscription.userId` for atomic sites, basically?

AllTerrainDeveloper replied:
Yes

fredrikekelund reviewed (changes requested):
Definitely shaping up nicely 👍 I had a few nit comments and one potential suggestion in `useDomainOwnerUserName`.

We should also resolve [this thread](https://github.com/Automattic/wp-calypso/pull/68158#discussion_r981216680) before merging.

And I think we still need to remove `loginUrlWithUserNameAndRedirectToEmailProvidersComparison` 😁

AllTerrainDeveloper committed:
Emails: Code improvements, rel = external on login link, moved function queryString to a higher scope.

AllTerrainDeveloper committed:
Emails: Remove duplicated link

wongasy reviewed (changes requested):
@AllTerrainDeveloper I noticed an issue where I'm seeing the non-owner screen in the email upsell step for adding a domain:

<img width="901" alt="Screen Shot 2022-09-29 at 2 02 38 PM" src="https://user-images.githubusercontent.com/104910361/193108720-ba79645b-4e28-4a88-8850-24b8e39410cd.png">

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,134 @@
+		if ( ownerUserName ) {
+			reasonText = translate(
+				'Email service can only be purchased by %(ownerUserName)s, ' +
+					'who is the owner of %(selectedDomainName)s. ' +
```
Can we bold the username as well as the domain name so they stand out more? The username within the hyperlink can be left as is.

AllTerrainDeveloper committed:
Emails: Fix domain purchase flow, add strong tag to domain and user name

AllTerrainDeveloper commented:
> @AllTerrainDeveloper I noticed an issue where I'm seeing the non-owner screen in the email upsell step for adding a domain:
> 
> <img alt="Screen Shot 2022-09-29 at 2 02 38 PM" width="901" src="https://user-images.githubusercontent.com/104910361/193108720-ba79645b-4e28-4a88-8850-24b8e39410cd.png">

Good catch, fixed.

AllTerrainDeveloper committed:
Emails: Remove un-needed properties, as well as translations not needed

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,117 @@
+				<p>{ reasonText }</p>
+			</PromoCard>
+			<TrackComponentView
+				eventName="calypso_non_domain_owner_notice_component_impression"
```
We're tracking the click as `calypso_email_providers_nonowner_click`, let's keep the naming conversion consistent :) 

```suggestion
				eventName="calypso_email_providers_nonowner_impression"
```

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,117 @@
+			</PromoCard>
+			<TrackComponentView
+				eventName="calypso_non_domain_owner_notice_component_impression"
+				eventProperties={ { source: 'email-comparison' } }
```
We could provide additional context on the reason why user got this message:

```suggestion
				eventProperties={ { source: 'email-comparison', context: 'non-domain-owner' } }
```

wongasy replied:
```suggestion
				eventProperties={ { source } }
```
We can dynamically report the source. As this notice may show up in the Email Management screen as well.

wongasy reviewed (approved):
@AllTerrainDeveloper LGTM! Just two minor comments around tracking.

The messaging shows up as expected for non-domain-owner:

<img width="1019" alt="Screen Shot 2022-09-30 at 11 09 33 AM" src="https://user-images.githubusercontent.com/104910361/193300660-8d7b3686-565a-4424-82d9-67667b0b682c.png">

The Email Comparison screen shows up as expected in the "Add a domain" flow:

<img width="1033" alt="Screen Shot 2022-09-30 at 11 10 23 AM" src="https://user-images.githubusercontent.com/104910361/193300803-1f96dccd-6dd1-4b7f-bfb1-70468e1ef415.png">

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7557365

Thank you @AllTerrainDeveloper for including a screenshot in the description! This is really helpful for our translators.

wongasy commented:
@AllTerrainDeveloper Other than the minor comment I left regarding tracking, I also found a weird issue, where when user gets redirected to the email comparison page, the Professional Email card is left collapsed. I think `source=login-redirect` might be triggering that logic for collapsing the PE card.

<img width="1396" alt="Screen Shot 2022-09-30 at 1 11 52 PM" src="https://user-images.githubusercontent.com/104910361/193322487-83b344d4-f7e1-4554-b1b8-3f42fdb2415e.png">

wongasy commented on client/my-sites/email/email-non-domain-owner-message/index.tsx:
```diff
@@ -0,0 +1,117 @@
+
+	const onClickLink = ( eventType: 'contact' | 'login' | 'support' ) => {
+		const properties = {
+			action: eventType,
```
```suggestion
			action: eventType,
			context: 'non-domain-owner'
```

Let's include the context here, so we can differentiate the clicks and impressions from non-domain-owner and non-email-owner.

wongasy replied:
```suggestion
			action: eventType,
			source,
```

Let's pick up the source and track it in the click as well!

AllTerrainDeveloper commented:
That issue was a tough one! Due to the nature of the async properties being fetched, the initial state is triggered with the initial loading state (which **may** or **not** have finished fetching everything)

I've added a `useEffect` hook with a dependency to the flag that is taking time to load (and will also determine the visibility of the notice)

AllTerrainDeveloper committed:
Add source as a parameter for Tracks in view component and Track clicks.
Fix async issue happening due to the prop that handles the notice visibility sometimes not fetched yet.

AllTerrainDeveloper committed:
Emails: Change event name to calypso_email_providers_nonowner_impression

wongasy commented:
@AllTerrainDeveloper I ran into an issue while reviewing the [other PR](https://github.com/Automattic/wp-calypso/pull/68435) on the Email Management page changes. 

When we remove all mailboxes, the non-owner will see the Email Comparison screen instead of the Email Management screen with the warning messages. This becomes an issue because non-owner admins could end up setting up Email Forwarding, when we have an active email subscription in place. The ideal fix would be to ensure non-owner users land on the Email Management page with the appropriate warning. However, that may involve a move complex fix with routing. As an alternative, we can simply update the Email Comparison screen to hide the Email Forwarding for non-owners.

<img width="1124" alt="Screen Shot 2022-10-03 at 2 36 41 PM" src="https://user-images.githubusercontent.com/104910361/193654465-a6de1a4d-96f3-4e6b-8080-c8df7961a0ab.png">

AllTerrainDeveloper committed:
Emails: Remove the Forward product link at the bottom of the screen when the user is not an owner.

wongasy commented:
@AllTerrainDeveloper Confirming that the email forward link has been hidden for non-owners.

<img width="1044" alt="Screen Shot 2022-10-04 at 12 24 58 PM" src="https://user-images.githubusercontent.com/104910361/193873570-544b883e-87d8-4b48-8dd2-e5a41d454a90.png">

fredrikekelund commented:
> When we remove all mailboxes, the non-owner will see the Email Comparison screen instead of the Email Management screen with the warning messages

I wasn't able to reproduce this. Here's what I did:

- On a site with a single domain with Professional Email.
- With the owner account, delete all mailboxes in that subscription.
- With the non-owner account, go to Upgrades > Emails in Calypso.
- I am immediately forwarded to the `titan/set-up-mailbox` page.

Could you outline the steps to reproduce this @wongasy?

Also, with this discussion in mind pbLl1t-1z0-p2, I'm thinking that we maybe shouldn't hide the email forwarding option for non-owner users? Being able to set up email forwarding if there's an existing email subscription sounds like a bug, but right now it looks like non-owners will never see that `Start with Email Forwarding` link, right?

wongasy commented:
>Could you outline the steps to reproduce this?
@fredrikekelund Hm... looks like I'm also unable to reproduce this. Initially I thought this could be related to the sandbox store, but I've tried both the regular store and sandbox store, and I was not able to reproduce the issue, which is great!

>I'm thinking that we maybe shouldn't hide the email forwarding option for non-owner users? Being able to set up email forwarding if there's an existing email subscription sounds like a bug, but right now it looks like non-owners will never see that Start with Email Forwarding link, right?
If we allow any non-owners to do the initial enabling of email forwarding, then there's a lower chance for the domain-owner to see the Email Comparison page. When they visit Upgrades > Email, they'll see the Email Forwarding management options with only a small button that takes them to the Email Comparison page. From the perspective of promoting a paid service, I think it makes sense to hide the email forwarding option for non-owners. Essentially, only the domain-owner should have the right to choose an email solution (paid or free). And once the email solution is chosen, then non-owner admins can manage the solution, in the case of email forward, they can set up additional forwarding (what we've [fixed](https://github.com/Automattic/wp-calypso/pull/68491) recently). What do you think?

wongasy committed:
Show email forwarding link for non-domain-owner admins

wongasy commented:
I do agree that the link for setting up Email Forwarding is so subtle that the potential conversion impact may not actually be a valid concern. In the interest of wrapping this up quickly, @AllTerrainDeveloper I've gone ahead and reverted the change for hiding the Email Forwarding Link.

AllTerrainDeveloper committed:
Add context to Tracks

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7557365).
